### PR TITLE
Grid fitting for RGB plot. Check for the latest version.

### PR DESCRIPTION
**Pull request by dmgav**
The features implemented in this PR:

- Automated check for the new version of PyXRF at nsls2forge. The message about new version availability and the new version number is displayed as part of the Main Window title.

- Implemented grid interpolation for RGB plot.

### CODE REVIEW

**Comment by dmgav** [comment]
Additional feature added: checkbox that enables uniform grid interpolation for data plotted vs. pixel numbers and x-y coordinates.

File: pyxrf/model/draw_image_rgb.py [hunk 1]
```diff
@@ -290,17 +309,69 @@ def _compute_equal_axes_ranges(x_min, x_max, y_min, y_max):
...
 
-        # Set equal ranges for the axes data
-        yd, xd = selected_data.shape[1], selected_data.shape[2]
-        xd_min, xd_max, yd_min, yd_max = 0, xd, 0, yd
-        # Select minimum range for data
-        if (yd <= math.floor(xd / 100)) and (xd >= 200):
-            yd_min, yd_max = -math.floor(xd / 200), math.ceil(xd / 200)
-        if (xd <= math.floor(yd / 100)) and (yd >= 200):
-            xd_min, xd_max = -math.floor(yd / 200), math.ceil(yd / 200)
+        def _adjust_data_range__min_ratio(c_min, c_max, c_axis_range, *, min_ratio=0.01):
```
Comment (mrakitin) [inline pyxrf/model/draw_image_rgb.py]: I don't think the double underscore in the name is in any naming convention.
```suggestion
        def _adjust_data_range_min_ratio(c_min, c_max, c_axis_range, *, min_ratio=0.01):
```

File: pyxrf/model/draw_image_rgb.py [hunk 2]
```diff
@@ -290,17 +309,69 @@ def _compute_equal_axes_ranges(x_min, x_max, y_min, y_max):
...
+            """
+            Adjust the range for plotted data along one axis (x or y). The adjusted range is
+            applied to the 'extend' attribute of imshow(). The adjusted range is always greater
```
Comment (mrakitin) [inline pyxrf/model/draw_image_rgb.py]: https://matplotlib.org/3.1.0/api/_as_gen/matplotlib.pyplot.imshow.html
```suggestion
            applied to the 'extent' attribute of imshow(). The adjusted range is always greater
```

File: pyxrf/model/draw_image_rgb.py [hunk 3]
```diff
@@ -290,17 +309,69 @@ def _compute_equal_axes_ranges(x_min, x_max, y_min, y_max):
...
+
+            # xd_min, xd_max, yd_min, yd_max = min(self.x_pos), max(self.x_pos),
+            #     min(self.y_pos), max(self.y_pos)
+            x_pos_2D = self.data_dict['positions']['x_pos']
+            y_pos_2D = self.data_dict['positions']['y_pos']
+            xd_min, xd_max, yd_min, yd_max = x_pos_2D.min(), x_pos_2D.max(), y_pos_2D.min(), y_pos_2D.max()
+            xd_axis_min, xd_axis_max, yd_axis_min, yd_axis_max = \
+                _compute_equal_axes_ranges(xd_min, xd_max, yd_min, yd_max)
+
+            xd_min, xd_max = _adjust_data_range__min_ratio(xd_min, xd_max, xd_axis_max - xd_axis_min)
```
Comment (mrakitin) [inline pyxrf/model/draw_image_rgb.py]:
```suggestion
            xd_min, xd_max = _adjust_data_range_min_ratio(xd_min, xd_max, xd_axis_max - xd_axis_min)
```

File: pyxrf/model/draw_image_rgb.py [hunk 4]
```diff
@@ -290,17 +309,69 @@ def _compute_equal_axes_ranges(x_min, x_max, y_min, y_max):
...
+            yd_min, yd_max = _adjust_data_range__min_ratio(yd_min, yd_max, yd_axis_max - yd_axis_min)
```
Comment (mrakitin) [inline pyxrf/model/draw_image_rgb.py]:
```suggestion
            yd_min, yd_max = _adjust_data_range_min_ratio(yd_min, yd_max, yd_axis_max - yd_axis_min)
```

File: pyxrf/model/draw_image_rgb.py [hunk 5]
```diff
@@ -290,17 +309,69 @@ def _compute_equal_axes_ranges(x_min, x_max, y_min, y_max):
...
+            #   during the experiment. Data is plotted starting from the upper-right corner of the plot
+            if x_pos_2D[0, 0] > x_pos_2D[0, -1]:
+                xd_min, xd_max, xd_axis_min, xd_axis_max = xd_max, xd_min, xd_axis_max, xd_axis_min
+            if y_pos_2D[0, 0] > y_pos_2D[-1, 0]:
+                yd_min, yd_max, yd_axis_min, yd_axis_max = yd_max, yd_min, yd_axis_max, yd_axis_min
+
+        else:
+
+            # Set equal ranges for the axes data
+            yd, xd = selected_data.shape[1], selected_data.shape[2]
```
Comment (mrakitin) [inline pyxrf/model/draw_image_rgb.py]: Is it a 3d array? What's the 0th dimension?

File: pyxrf/model/fileio.py [hunk 6]
```diff
@@ -88,12 +88,62 @@ def __init__(self, **kwargs):
         self.mask_data = None
 
         # Display PyXRF version in the window title
-        global pyxrf_version
-        if pyxrf_version[0] not in 'vV':
-            pyxrf_version = f"v{pyxrf_version}"
-        self.window_title_base = f"PyXRF: X-ray Fluorescence Analysis Tool ({pyxrf_version})"
+        ver, new_ver = self._get_pyxrf_version_str()
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]: Maybe name the `ver` as `current_ver` here and below?

File: pyxrf/model/fileio.py [hunk 7]
```diff
@@ -88,12 +88,62 @@ def __init__(self, **kwargs):
...
+        if new_ver:
+            ver_str += f" - new version {new_ver} is available"
+        self.window_title_base = f"PyXRF: X-ray Fluorescence Analysis Tool ({ver_str})"
         self.window_title = self.window_title_base
 
+    def _get_pyxrf_version_str(self):
+        """
+        The function returns the tuple of strings:
+        - current version number of PyXRF;
+        - the latest version of PyXRF from nsls2forge channel.
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]:
```suggestion
        - the latest version of PyXRF from nsls2forge conda channel.
```

File: pyxrf/model/fileio.py [hunk 8]
```diff
@@ -88,12 +88,62 @@ def __init__(self, **kwargs):
...
+        If current version is the latest, then the second string is empty
+        """
+
+        import requests
+        from distutils.version import LooseVersion
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]: I'd recommend to move those imports to the beginning of the file.

File: pyxrf/model/fileio.py [hunk 9]
```diff
@@ -88,12 +88,62 @@ def __init__(self, **kwargs):
...
+
+        # Determine the current version of PyXRF
+        global pyxrf_version
+        pyxrf_version_str = pyxrf_version
+        if pyxrf_version_str[0] not in 'vV':
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]:
```suggestion
        if not pyxrf_version_str.lower().startswith('v'):
```

File: pyxrf/model/fileio.py [hunk 10]
```diff
@@ -88,12 +88,62 @@ def __init__(self, **kwargs):
...
+            pkgs = r.json()
+            pyxrf_ver = []
+            for pkg in pkgs['packages'].keys():
+                if pkg.startswith('pyxrf'):
+                    pkg_version = LooseVersion(pkg.split('-')[1])
+                    pyxrf_ver.append(pkg_version)
+
+            if len(pyxrf_ver):
+                max_version = pyxrf_ver[0]
+                for n, pkg_version in enumerate(pyxrf_ver):
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]: Why do we need `enumerate()` and `n` here? They don't seem to be used...

File: pyxrf/model/fileio.py [hunk 11]
```diff
@@ -88,12 +88,62 @@ def __init__(self, **kwargs):
...
+                current_version = LooseVersion(pyxrf_version)
+
+                if current_version < max_version:
+                    pyxrf_latest_version_str = f"v{max_version}"
+
+        except Exception:
+            # This exception is mostly likely to happen if there is no internet connection or
+            #    nsls2forge is unreachable. Then ignore the procedure, assume that the current
+            #    version is the latest.
+            logger.info("Failed to check availability of the latest version of PyXRF at 'nsls2forge'.")
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]:
```suggestion
            logger.info("Failed to check availability of the latest version of PyXRF in the 'nsls2forge' conda channel.")
```

File: pyxrf/model/fileio.py [hunk 12]
```diff
@@ -88,12 +88,62 @@ def __init__(self, **kwargs):
...
+        from distutils.version import LooseVersion
+
+        # Determine the current version of PyXRF
+        global pyxrf_version
+        pyxrf_version_str = pyxrf_version
+        if pyxrf_version_str[0] not in 'vV':
+            pyxrf_version_str = f"v{pyxrf_version_str}"
+
+        # Now find the latest version available at nsls2forge
+        pyxrf_latest_version_str = ""
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]: Will it make sense to use `None` for that if there is no new version? As you don't concatenate strings here, I think it's OK. If so, the docs above should be updated.
```suggestion
        pyxrf_latest_version_str = None
```

File: pyxrf/model/draw_image_rgb.py (hunk 1 above)
Comment (dmgav) [reply pyxrf/model/draw_image_rgb.py]: Do you think it's a violation of naming convention? Typically naming convention restricts the number of underscores at the beginning of the name. I rarely use double underscores to separate two parts of the name. If you insist on changing, I guess the right name will be `adjust_data_range_using_min_ratio` or something like this, but it becomes long.

File: pyxrf/model/draw_image_rgb.py (hunk 4 above)
Comment (dmgav) [reply pyxrf/model/draw_image_rgb.py]: Same question.

File: pyxrf/model/draw_image_rgb.py (hunk 5 above)
Comment (dmgav) [reply pyxrf/model/draw_image_rgb.py]: Yes, it is 3d array. The 0th dimension is the number of the image (RGB color):

Lines 376-381 of the same file:

```
        name_r = self.rgb_name_list[self.index_red]
        data_r = selected_data[self.index_red, :, :]
        name_g = self.rgb_name_list[self.index_green]
        data_g = selected_data[self.index_green, :, :]
        name_b = self.rgb_name_list[self.index_blue]
        data_b = selected_data[self.index_blue, :, :]
```

File: pyxrf/model/fileio.py (hunk 8 above)
Comment (dmgav) [reply pyxrf/model/fileio.py]: I was planning to move it, but forgot.

File: pyxrf/model/fileio.py (hunk 10 above)
Comment (dmgav) [reply pyxrf/model/fileio.py]: Yes, I forgot to remove it. `enumerate` was used in debugging.

File: pyxrf/model/fileio.py (hunk 12 above)
Comment (dmgav) [reply pyxrf/model/fileio.py]: You may be right. Initially this function was supposed to return complete formatted string. In current implementation I think using using None is fine.

File: pyxrf/model/fileio.py (hunk 9 above)
Comment (dmgav) [reply pyxrf/model/fileio.py]: I think that the first version is more computationally efficient, but the efficiency doesn't really matter here.

File: pyxrf/gui.py [hunk 13]
```diff
@@ -69,16 +78,10 @@ def run():
...
-    formatter = logging.Formatter(fmt='%(asctime)s : %(levelname)s : %(message)s')
     guihandler = GuiHandler()
     guihandler.setLevel(logging.INFO)
     guihandler.setFormatter(formatter)
-    stream_handler = logging.StreamHandler()
-    stream_handler.setFormatter(formatter)
-    stream_handler.setLevel(logging.INFO)
-    logger.setLevel(logging.INFO)
     logger.addHandler(guihandler)
-    logger.addHandler(stream_handler)
```
Comment (mrakitin) [inline pyxrf/gui.py]: +1 for separating those 2 handlers.

File: pyxrf/model/draw_image_rgb.py (hunk 5 above)
Comment (mrakitin) [reply pyxrf/model/draw_image_rgb.py]: OK, thanks for the explanation.

File: pyxrf/model/fileio.py (hunk 6 above)
Comment (mrakitin) [reply pyxrf/model/fileio.py]: What about this?

File: pyxrf/model/fileio.py (hunk 7 above)
Comment (mrakitin) [reply pyxrf/model/fileio.py]: This suggestion was left uncommitted...

File: pyxrf/model/fileio.py [hunk 14]
```diff
@@ -88,12 +91,68 @@ def __init__(self, **kwargs):
...
+        The function returns the tuple of strings:
+        - current version number of PyXRF;
+        - the latest version of PyXRF from nsls2forge channel.
+        If current version is the latest, then the second string None.
+        """
+
+        # Determine the current version of PyXRF
+        global pyxrf_version
+        pyxrf_version_str = pyxrf_version
+        if not pyxrf_version_str.lower().startswith('v'):
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]: Can be a combination of your original implementation and my suggestion if we want a better efficiency :), such as:
```suggestion
        if pyxrf_version_str[0].lower() != 'v':
```

File: pyxrf/model/fileio.py (hunk 10 above)
Comment (mrakitin) [reply pyxrf/model/fileio.py]: It's still here. Do you mind to remove it?

File: pyxrf/model/fileio.py [hunk 15]
```diff
@@ -88,12 +91,68 @@ def __init__(self, **kwargs):
...
+                        max_version = pkg_version
+
+                current_version = LooseVersion(pyxrf_version)
+
+                if current_version < max_version:
+                    pyxrf_latest_version_str = f"v{max_version}"
+
+            if pyxrf_latest_version_str is not None:
+                logger.info(f"New version of PyXRF ({pyxrf_latest_version_str}) was found "
+                            "at the 'nsls2forge' conda channel")
```
Comment (mrakitin) [inline pyxrf/model/fileio.py]: Maybe use `in` to be consistent with the message in the `except`?
```suggestion
                            "in the 'nsls2forge' conda channel")
```

File: pyxrf/gui.py (hunk 13 above)
Comment (dmgav) [reply pyxrf/gui.py]: The handler was added after GUI initialization, so messages was printed in the terminal.

**Comment by dmgav** [comment]
I think I committed everything. At least I see all the changes in the actual file.

**Pull request merged**